src/asammdf/mdf.py

```diff
@@ -2528,9 +2971,11 @@ def concatenate(
 
                     # Make a channel group translation dictionary if the order is different
                     if make_translation:
-                        for i, org_group in enumerate(first_mdf.groups):
+                        first_mdf._mdf = typing.cast(mdf_v4.MDF4, first_mdf._mdf)
+                        mdf._mdf = typing.cast(mdf_v4.MDF4, mdf._mdf)
+                        for i, org_group in enumerate(first_mdf._mdf.groups):
                             org_group_source = org_group.channel_group.acq_source
-                            for j, new_group in enumerate(mdf.groups):
+                            for j, new_group in enumerate(mdf._mdf.groups):
```

Comment:
I guess you can use the property `groups` directly here

```python
for j, new_group in enumerate(mdf.groups):
```

Reply:
On this line, a list of GroupV4 (MDF4) is expected so I am getting this from `_mdf`  because `groups` might either be of type GroupV3 (for MDF3) or GroupV4 (for MDF4) whereas `_mdf.groups` is assured to be of type GroupV4 since `_mdf` is of type MDF4